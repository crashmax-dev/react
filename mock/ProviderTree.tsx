import React from 'react'

export function App() {
  return <div>This is App</div>
}

type MockProviderProps = React.PropsWithChildren<
  React.HTMLAttributes<HTMLDivElement>
>

export function MockProvider({ children, ...props }: MockProviderProps) {
  return <div {...props}>{children}</div>
}
