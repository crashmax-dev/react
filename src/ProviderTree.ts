import { createElement } from 'react'

type Component = React.FunctionComponent<any>
type Props<T extends Component> = Omit<Parameters<T>[0], 'children'>
type Provider<T extends Component> = [Component, Props<T>?]

export function ProviderTree(props: {
  providers: (
    wrapper: <T extends Component>(
      component: T,
      props?: Props<T>
    ) => Provider<T>
  ) => Provider<Component>[]
  children?: JSX.Element
}) {
  const providers = props.providers(providerWrapper)
  return createProviderTree(providers, props.children)
}

function createProviderTree<T extends Component>(
  providers: Provider<T>[],
  children?: JSX.Element
): JSX.Element {
  const [provider, props = {}] = providers.shift()!
  return createElement(
    provider,
    props,
    providers.length ? createProviderTree(providers, children) : children
  )
}

function providerWrapper<T extends Component>(
  provider: T,
  props?: Parameters<T>[0]
): Provider<T> {
  return [provider, props]
}
