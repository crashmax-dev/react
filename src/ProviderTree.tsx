import React from 'react'
import type { JSXElementConstructor, ReactNode } from 'react'

type NoInfer<T> = [T][T extends any ? 0 : 1]

type ContainsChildren = {
  children?: React.ReactNode
}

/**
 * Creates a provider tree with the given providers and children.
 *
 * @param {function} props.providers
 * A function that returns an array of component providers.
 *
 * @param {ReactNode} [props.children]
 * The children to render inside the provider tree.
 *
 * @throws {Error}
 * When there are no providers provided.
 *
 * @return {ReactNode}
 * The provider tree with the given providers and children.
 *
 * @example
 * ```jsx
 * <ProviderStack
 *   providers={[
 *     [StrictMode],
 *     [ThemeProvider, { theme: 'dark' }]
 *   ]}
 * >
 *   <App />
 * </ProviderStack>
 * ```
 */
export function ProviderTree<
  Providers extends [ContainsChildren, ...ContainsChildren[]]
>({
  providers,
  children
}: {
  providers: {
    [K in keyof Providers]: [
      provider: JSXElementConstructor<Providers[K]>,
      props?: Omit<NoInfer<Providers[K]>, 'children'>
    ]
  }
  children: ReactNode
}): ReactNode {
  let node = children

  for (let i = providers.length - 1; i >= 0; i--) {
    const [Provider, props] = providers[i]!
    node = <Provider {...props}>{node}</Provider>
  }

  return node
}
