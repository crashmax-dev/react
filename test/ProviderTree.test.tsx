import { App, MockProvider } from '../mock/ProviderTree.js'
import { ProviderTree } from '../src/ProviderTree.js'
import { render } from './utils.js'

describe('ProviderTree', () => {
  it('should be defined', () => {
    expect(ProviderTree).toBeDefined()
  })

  it('should render the tree of providers', () => {
    const { container } = render(
      <ProviderTree
        providers={($) => [
          $(MockProvider, { id: 'foo' }),
          $(MockProvider, { id: 'bar' })
        ]}
      >
        <App />
      </ProviderTree>
    )
    expect(container).toMatchSnapshot()
  })

  it('should render the tree of providers without props', () => {
    const { container } = render(
      ProviderTree({
        providers: ($) => [$(MockProvider), $(MockProvider)]
      })
    )

    expect(container).toMatchSnapshot()
  })

  it('should render the tree of providers with a single provider', () => {
    const { container } = render(
      ProviderTree({
        providers: ($) => [$(MockProvider, { id: 'single' })]
      })
    )

    expect(container).toMatchSnapshot()
  })

  it('should be expect error when no providers are provided', () => {
    expect(() =>
      render(
        ProviderTree({
          providers: ($) => []
        })
      )
    ).toThrowErrorMatchingSnapshot()
  })
})
