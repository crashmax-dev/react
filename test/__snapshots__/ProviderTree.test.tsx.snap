// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProviderTree > should render the tree of providers 1`] = `
<div>
  <div
    id="foo"
  >
    <div
      id="bar"
    >
      <div>
        This is App
      </div>
    </div>
  </div>
</div>
`;

exports[`ProviderTree > should render the tree of providers with a single provider 1`] = `
<div>
  <div
    id="single"
  />
</div>
`;

exports[`ProviderTree > should render the tree of providers without props 1`] = `
<div>
  <div>
    <div />
  </div>
</div>
`;
